processing/src/main/java/io/druid/query/groupby/epinephelinae/ConcurrentGrouper.java

```diff
@@ -285,26 +308,62 @@ private boolean isParallelSortAvailable()
     }
   }
 
-  private List<Iterator<Entry<KeyType>>> getGroupersIterator(boolean sorted)
+  private List<CloseableIterator<Entry<KeyType>>> getGroupersIterator(boolean sorted)
   {
     return groupers.stream()
                    .map(grouper -> grouper.iterator(sorted))
                    .collect(Collectors.toList());
   }
 
+  /**
+   * Merge dictionaries of {@link Grouper.KeySerde}s of {@link Grouper}s.  The result dictionary contains unique string
+   * keys.
+   *
+   * @return merged dictionary if its size does not exceed max dictionary size.  Otherwise null.
+   */
+  @Nullable
+  private List<String> tryMergeDictionary()
+  {
+    final Set<String> mergedDictionary = new HashSet<>();
+    long totalDictionarySize = 0L;
+
+    for (SpillingGrouper<KeyType> grouper : groupers) {
+      final List<String> dictionary = grouper.getDictionary();
+
+      for (String key : dictionary) {
+        if (!mergedDictionary.contains(key)) {
+          totalDictionarySize += RowBasedGrouperHelper.estimateStringKeySize(key);
+          if (totalDictionarySize > maxDictionarySizeForCombiner) {
+            return null;
+          }
+          mergedDictionary.add(key);
+        }
+      }
+    }
+
+    return ImmutableList.copyOf(mergedDictionary);
```

Comment:
I'm not sure. The dictionary should be a list in RowBasedKeySerde, so it should be converted at some point if this returns a collection.
The dictionary in RowBasedKeySerde is basically mutable for run-time dictionary generation. To avoid too many duplicated codes, I used the same logic here except that the dictionary is generated first and shared by all threads. The reason for returning an immutable list here is for the safety.
I think we can do better by refactoring RowBasedKeySerde somehow, but should be done in a follow-up pr.

Reply:
I read `RowBasedKeySerde` and didn't find list is really required there. Seems that you just iterate it with index. You can do the same with collection, for-each iteration on it + increment index in parallel.

Reply:
Please check AbstractStringRowBasedKeySerdeHelper. ```dictionary``` is randomly accessed in bufferComparator and getFromByteBuffer().